### PR TITLE
feat(dynamics): make convergence migration compile-safe

### PR DESCRIPTION
**Pull request by laserkelvin**
# ALCHEMI Toolkit Pull Request

## Description

Make convergence evaluation and status migration compile-safe so multi-stage pipelines such as FIRE2 to NVT can keep the convergence machinery inside the compiled region.

The previous path performed host-synchronizing, data-dependent operations on every step: `evaluate()` gated on `converged_mask.any()` and returned dynamically shaped `torch.where(mask)[0]` indices, migration used boolean fancy indexing behind another `.any()`, and the fused step conditionally dispatched `ON_CONVERGE` hooks based on a host-evaluated tensor. These operations introduce graph breaks or per-step GPU-to-CPU stalls.

This PR is stacked on #168. Until that PR merges, GitHub will show the parent compile-step commits in this PR as well.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Depends on #168.

## Changes Made

- Add `ConvergenceHook.evaluate_mask()`, a branchless core that returns a static-shape boolean `(B,)` mask without host synchronization, while retaining `evaluate()` as the index-returning wrapper for uncompiled callers.
- Blend migrated statuses with `torch.where` instead of `.any()` gating and boolean fancy indexing.
- Store per-substage convergence masks in `_last_converged` and dispatch registered `ON_CONVERGE` hooks unconditionally with the current mask in `ctx.converged_mask`.
- Add a static hook-stage lookup used to decide whether an `ON_CONVERGE` hook is registered without inspecting tensor data.
- Update fused-stage regression tests for all-false masks and stage-local active masks.

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Focused verification performed for this PR:

```text
uv run --extra cu13 pytest --no-cov \
  test/dynamics/test_single_loop.py \
  test/dynamics/test_base.py \
  test/hooks/test_registry.py
# 189 passed

uv run --extra cu13 ruff check \
  nvalchemi/dynamics/base.py \
  nvalchemi/hooks/_registry.py \
  test/dynamics/test_single_loop.py
# All checks passed

uv run --extra cu13 ruff format --check \
  nvalchemi/dynamics/base.py \
  nvalchemi/hooks/_registry.py \
  test/dynamics/test_single_loop.py
# 3 files already formatted
```

The broader compiled FIRE2 to NVT benchmark measured 7 graph breaks, matching single-stage NVT, with convergence and migration contributing zero breaks. The compiled default run measured 14.0 ms/step versus 21.8 ms/step eager. Those benchmark measurements were not rerun during this publishing pass.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

The user-visible contract change is that registered fused-stage `ON_CONVERGE` hooks now run every step and must consult `ctx.converged_mask`; an all-false mask means no samples converged during that step. The ordinary uncompiled `BaseDynamics` path retains its existing conditional callback behavior.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Review by dallasfoster — COMMENTED** [review]
Can you show proof that the net, per-step sync counts go down for a FIRE2 -> NVT compiled workflow, with a ConvergedSnapshotHook hook attached?

**Comment by laserkelvin** [comment]
@dallasfoster Measured exactly that configuration: compiled FIRE2 → NVT `FusedStage` (composed via `fire + nvt`) with a `ConvergedSnapshotHook` writing converged systems to a `GPUBuffer` sink. Per-step host syncs are counted with `torch.cuda.set_sync_debug_mode("warn")` around each `step()` + `all_complete` check (mirroring `run()`), after warmup/compile.

| Pin | Mean syncs/step | FIRE phase | NVT phase | Total (59 steps) |
|---|---:|---:|---:|---:|
| upstream `main` (pre-#168) | 28.17 | 27.51 | 29.45 | 1662 |
| PR base (#168 head, `7b02ac6`) | 8.49 | 7.18 | 11.05 | 501 |
| **This PR (`c09fe25`)** | **3.69** | **3.72** | **3.65** | **218** |

All three runs are physically identical (same 59-step trajectory, same convergence step, same snapshot writes). Two things worth noting: on the base, the sync count *grows* once systems migrate to NVT — the emptied FIRE stage's gated convergence checks keep syncing every step — while the mask-based path stays flat at ~3.7/step. And that flat number holds even though `ConvergedSnapshotHook` now fires every step under the new contract: its only host decision is the single `.any()` inside its own `torch.compiler.disable` island.

The MWE below is self-contained — `uv run exp13_sync_mwe.py` reproduces the *before* row in an isolated environment (the inline metadata pins the PR base commit); swap the pin to this PR's head for the *after* row. Numbers above were produced from exactly these isolated runs.

```python
# /// script
# requires-python = ">=3.11,<3.14"
# dependencies = [
#   "nvalchemi-toolkit[cu13] @ git+https://github.com/laserkelvin/nvalchemi-toolkit@7b02ac6956b7a3c30db4fe6ca2f103f951c153b5",
# ]
# ///
"""MWE: per-step CUDA sync counts for a compiled FIRE2 -> NVT pipeline.

Answers the PR #169 review question: do net per-step host synchronizations
go down with mask-based convergence, for a compiled FIRE2 -> NVT
``FusedStage`` with a ``ConvergedSnapshotHook`` attached?

``torch.cuda.set_sync_debug_mode("warn")`` makes every synchronizing torch
call emit a ``UserWarning``; the loop records them per step (one ``step()``
plus the ``all_complete`` termination check, mirroring ``run()``).

The inline metadata pins ``nvalchemi-toolkit`` to this PR's base commit
(the PR #168 head), so ``uv run exp13_sync_mwe.py`` reproduces the
*before* numbers in isolation. For the *after* numbers, swap the pin to
the PR head commit. (Pinning upstream ``main`` instead measures the
pre-#168 baseline — higher still.)
"""

from __future__ import annotations

import json
import warnings
from collections import Counter

import torch

import nvalchemi
from nvalchemi.data import AtomicData, Batch
from nvalchemi.dynamics import NVTLangevin
from nvalchemi.dynamics.base import ConvergenceHook, DynamicsStage, FusedStage
from nvalchemi.dynamics.hooks import ConvergedSnapshotHook
from nvalchemi.dynamics.optimizers import FIRE2
from nvalchemi.dynamics.sinks import GPUBuffer
from nvalchemi.hooks import NeighborListHook
from nvalchemi.models.lj import LennardJonesModelWrapper

N_SYSTEMS = 4
N_ATOMS = 64
MD_STEPS = 20
MAX_STEPS = 300

# Batch: rattled 4x4x4 argon grids at the LJ minimum spacing — real
# optimization work, but reliably converges so ON_CONVERGE actually fires.
torch.manual_seed(0)
spacing = 2 ** (1 / 6) * 3.40 * 1.05
coords = torch.arange(4, dtype=torch.float32) * spacing
gx, gy, gz = torch.meshgrid(coords, coords, coords, indexing="ij")
lattice = torch.stack([gx.flatten(), gy.flatten(), gz.flatten()], dim=-1)
batch = Batch.from_data_list(
    [
        AtomicData(
            positions=lattice + 0.25 * torch.randn_like(lattice),
            atomic_numbers=torch.full((N_ATOMS,), 18, dtype=torch.long),
            forces=torch.zeros(N_ATOMS, 3),
            energy=torch.zeros(1, 1),
            velocities=torch.zeros(N_ATOMS, 3),
        )
        for _ in range(N_SYSTEMS)
    ],
    device="cuda",
)

# Pipeline: FIRE2 optimization -> NVT MD, composed with the `+` sugar, with
# a ConvergedSnapshotHook persisting converged systems to a GPU sink.
model = LennardJonesModelWrapper(epsilon=0.0104, sigma=3.40, cutoff=8.5)
fire = FIRE2(
    model=model,
    dt=0.05,
    n_steps=MAX_STEPS,
    convergence_hook=ConvergenceHook.from_fmax(0.05),
)
sink = GPUBuffer(
    capacity=N_SYSTEMS,
    max_atoms=N_SYSTEMS * N_ATOMS,
    max_edges=N_SYSTEMS * N_ATOMS * 64,
    device="cuda",
)
fire.register_hook(ConvergedSnapshotHook(sink=sink))
nvt = NVTLangevin(
    model=model,
    dt=1.0,
    temperature=300.0,
    friction=0.05,
    n_steps=MD_STEPS,
    random_seed=7,
)
fused = fire + nvt
fused.register_hook(
    NeighborListHook(
        model.model_config.neighbor_config, stage=DynamicsStage.BEFORE_COMPUTE
    )
)

# Lazy init and compilation happen before counting starts.
for _ in range(2):
    batch, _ = fused.step(batch)
fused.compile()
for _ in range(3):
    batch, _ = fused.step(batch)

torch.cuda.set_sync_debug_mode("warn")
per_step: list[int] = []
status_trace: list[list[int]] = []
sync_ops: Counter[str] = Counter()
try:
    for _ in range(MAX_STEPS):
        with warnings.catch_warnings(record=True) as caught:
            warnings.simplefilter("always")
            batch, _ = fused.step(batch)
            done = FusedStage.all_complete(batch, fused.exit_status)
        syncs = [w for w in caught if "synchron" in str(w.message).lower()]
        per_step.append(len(syncs))
        for w in syncs:
            sync_ops[str(w.message)[:80]] += 1
        status_trace.append(batch.status.view(-1).tolist())
        if done:
            break
finally:
    torch.cuda.set_sync_debug_mode("default")

first_all_nvt = next(
    (i for i, st in enumerate(status_trace) if all(s >= 1 for s in st)),
    len(per_step),
)
fire_phase = per_step[:first_all_nvt] or [0]
nvt_phase = per_step[first_all_nvt:] or [0]
print(
    json.dumps(
        {
            "code": nvalchemi.__file__,
            "steps": len(per_step),
            "snapshots_written": len(sink),
            "syncs_per_step_mean": round(sum(per_step) / len(per_step), 2),
            "fire_phase_mean": round(sum(fire_phase) / len(fire_phase), 2),
            "nvt_phase_mean": round(sum(nvt_phase) / len(nvt_phase), 2),
            "total_syncs": sum(per_step),
            "top_sync_ops": sync_ops.most_common(5),
        },
        indent=2,
    )
)
```

**Comment by laserkelvin** [comment]
/ok to test [2a0db34](https://github.com/NVIDIA/nvalchemi-toolkit/pull/169/commits/2a0db340f756871d7a2c3003c41f26c93882f837)

**Review by dallasfoster — APPROVED** [review]

**Comment by dallasfoster** [comment]
/ok to test [e01fdc8](https://github.com/NVIDIA/nvalchemi-toolkit/pull/169/commits/e01fdc885874e111a6fa433432656687d51adfcb)
